src/rgw/rgw_notify.cc

```diff
@@ -2,20 +2,596 @@
+      ldout(cct, 5) << "WARNING: failed to create push endpoint: " 
+          << record_with_endpoint.push_endpoint << " for entry: " << entry.marker << ". error: " << e.what() << " (will retry) " << dendl;
+      return false;
+    }
+  }
+
+  bool queue_owned(const std::string& queue_name, spawn::yield_context yield) {
+    librados::ObjectWriteOperation op;
+    op.assert_exists();
+    rados::cls::lock::lock(&op, queue_name+"_lock", 
+        ClsLockType::EXCLUSIVE,
+        lock_cookie, 
+        "" /*no tag*/,
+        "" /*no description*/,
+        failover_time,
+        LOCK_FLAG_MUST_RENEW);
+
+    const auto ret = rgw_rados_operate(rados_ioctx, queue_name, &op, optional_yield(io_context, yield));
+    if (ret == -EBUSY) {
+      ldout(cct, 5) << "WARNING: queue: " << queue_name << " ownership moved to another daemon. processing will stop" << dendl;
+      return false;
+    }
+    if (ret == -ENOENT) {
+      ldout(cct, 10) << "INFO: check ownership of queue: " << queue_name << " - already deleted" << dendl;
+      return false;
+    }
+    if (ret < 0) {
+      ldout(cct, 1) << "ERROR: check ownership of queue: " << queue_name << ". error: " << ret << dendl;
+      return true; 
+    }
+    return true;
+  }
+
+  // clean stale reservation from queue
+  void cleanup_queue(const std::string& queue_name, spawn::yield_context yield) {
+    while (queue_owned(queue_name, yield)) {
+      ldout(cct, 20) << "INFO: trying to perform stale reservation cleanup for queue: " << queue_name << dendl;
+      const auto now = ceph::coarse_real_time::clock::now();
+      const auto stale_time = now - std::chrono::seconds(stale_reservations_period_s);
+      librados::ObjectWriteOperation op;
+      op.assert_exists();
+      cls_2pc_queue_expire_reservations(op, stale_time);
+      auto ret = rgw_rados_operate(rados_ioctx, queue_name, &op, optional_yield(io_context, yield));
+      if (ret == -ENOENT) {
+        // queue was deleted
+        ldout(cct, 5) << "INFO: queue: " 
+          << queue_name << ". was removed. cleanup will stop" << dendl;
+        return;
+      }
+      if (ret < 0) {
+        ldout(cct, 5) << "WARNING: failed to cleanup stale reservation from queue: " << queue_name
+          << ". error: " << ret << dendl;
+      }
+      Timer timer(io_context);
+      timer.expires_from_now(std::chrono::seconds(reservations_cleanup_period_s));
+      boost::system::error_code ec;
+	    timer.async_wait(yield[ec]);
+    }
+  }
+
+  // processing of a specific queue
+  void process_queue(const std::string& queue_name, spawn::yield_context yield) {
+    constexpr auto max_elements = 1024;
+    auto is_idle = false;
+    const std::string start_marker;
+
+    // start a the cleanup coroutine for the queue
+    spawn::spawn(io_context, [this, queue_name](spawn::yield_context yield) {
+            cleanup_queue(queue_name, yield);
+            });
+    
+    while (queue_owned(queue_name, yield)) {
```

Comment:
similar to cleanup_queue(), it seems like we're better off combining this with the existing osd ops. for the read op (cls_2pc_queue_list_entries) we can use assert_locked, and for the write op (cls_2pc_queue_remove_entries) we can renew

Reply:
agreed. see: https://github.com/ceph/ceph/pull/34293#discussion_r471998562